### PR TITLE
daemon,store: plug in authentication for store search/details

### PR DESCRIPTION
**Pull request by matiasb**
Get user from client request, if authorization was provided, and pass authenticator to the remote snap repository to make authenticated requests against the store.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -234,6 +234,9 @@ func loginUser(c *Command, r *http.Request) Response {
 // It requires the state to be locked
 func UserFromRequest(st *state.State, req *http.Request) (*auth.UserState, error) {
 	// extract macaroons data from request
+	if req == nil {
```
Comment (niemeyer) [inline daemon/api.go]: That feels like an error on the call site. We shouldn't call *FromRequest if we don't have a request.

File: daemon/api.go [hunk 2]
```diff
@@ -316,7 +319,20 @@ func getSnapInfo(c *Command, r *http.Request) Response {
 		channel = localSnap.Channel
 	}
 
-	remoteSnap, _ := remoteRepo.Snap(name, channel, nil)
+	var authenticator store.Authenticator
```
Comment (niemeyer) [inline daemon/api.go]: Can we use `auther` for the variable name, per prior conversation?

File: daemon/api.go (hunk 1 above)
Comment (chipaca) [inline daemon/api.go]: I don't know how often UserFromRequest is called, but I know that a lot of tests use a nil request. Maybe they should use a nullRequest instead, or something :-)

File: daemon/api.go [hunk 3]
```diff
@@ -316,7 +319,20 @@ func getSnapInfo(c *Command, r *http.Request) Response {
...
+	overlord := c.d.overlord
+	state := overlord.State()
+	state.Lock()
+	user, err := UserFromRequest(state, r)
+	state.Unlock()
+	if err != nil {
+		return InternalError("%v", err)
+	}
+	if user != nil {
```
Comment (niemeyer) [inline daemon/api.go]: If UserFromRequest returns err == nil, the user value should always be nil. We asked "give me a user" and it said "no problem". It's lying if the user is nil.

File: daemon/api.go [hunk 4]
```diff
@@ -413,13 +429,26 @@ func getSnapsInfo(c *Command, r *http.Request) Response {
 
 		remoteRepo := newRemoteRepo()
 
+		var authenticator store.Authenticator
```
Comment (niemeyer) [inline daemon/api.go]: auther

File: store/snap_remote_repo.go [hunk 5]
```diff
@@ -244,6 +244,9 @@ func (s *SnapUbuntuStoreRepository) Snap(name, channel string, auther Authentica
 	}
 
 	// set headers
+	if auther != nil {
+		auther.Authenticate(req)
+	}
 	s.applyUbuntuStoreHeaders(req, "")
```
Comment (pete-woods) [inline store/snap_remote_repo.go]: These should be factored out into a single method, something like:
func (s *SnapUbuntuStoreRepository) applyAuthorizedHeaders(req *http.Request, auther Authentication) {
  	s.applyUbuntuStoreHeaders(req, "")
	if auther != nil {
		auther.Authenticate(req)
	}
}

File: daemon/api_test.go [hunk 6]
```diff
@@ -303,6 +306,25 @@ func (s *apiSuite) TestSnapInfoOneIntegration(c *check.C) {
...
+	req, err := http.NewRequest("GET", "/v2/snap/foo", nil)
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Authorization", `Macaroon root="macaroon", discharge="discharge"`)
+
+	c.Assert(s.auther, check.IsNil)
+
+	_, ok := getSnapInfo(snapCmd, req).(*resp)
+	c.Assert(ok, check.Equals, true)
+	// ensure authenticator was set
+	c.Assert(s.auther, check.NotNil)
```
Comment (niemeyer) [inline daemon/api_test.go]: Can we check it always has the proper metadata instead, by type-asserting it to the right authenticator?

Comment (niemeyer) [inline daemon/api_test.go]: Actually, we don't even need type asserts, just check.DeepEquals.

File: store/snap_remote_repo.go [hunk 7]
```diff
@@ -244,6 +244,9 @@ func (s *SnapUbuntuStoreRepository) Snap(name, channel string, auther Authentica
 	}
 
 	// set headers
+	if auther != nil {
+		auther.Authenticate(req)
+	}
```
Comment (niemeyer) [inline store/snap_remote_repo.go]: Can we put that inside the function call below?

File: daemon/api.go (hunk 3 above)
Comment (matiasb) [inline daemon/api.go]: User can be nil if no authorization data is available in the request. In those cases, requests are assumed to be anonymous, and no authentication is set in the requests to the store, right?

**Comment by chipaca** [comment]
:+1:

File: daemon/api.go [hunk 8]
```diff
@@ -230,13 +231,17 @@ func loginUser(c *Command, r *http.Request) Response {
 	return SyncResponse(result, nil)
 }
 
+var (
```
Comment (niemeyer) [inline daemon/api.go]: Single line here please.

File: daemon/api.go [hunk 9]
```diff
@@ -316,7 +321,20 @@ func getSnapInfo(c *Command, r *http.Request) Response {
...
-	remoteSnap, _ := remoteRepo.Snap(name, channel, nil)
+	var auther store.Authenticator
+	overlord := c.d.overlord
+	state := overlord.State()
+	state.Lock()
+	user, err := UserFromRequest(state, r)
+	state.Unlock()
+	if err == nil {
+		auther = user.Authenticator()
+	}
```
Comment (niemeyer) [inline daemon/api.go]: `} else if err != errNoAuth {`

File: daemon/api.go [hunk 10]
```diff
@@ -413,13 +431,26 @@ func getSnapsInfo(c *Command, r *http.Request) Response {
...
+		state := overlord.State()
+		state.Lock()
+		user, err := UserFromRequest(state, r)
+		state.Unlock()
+		if err == nil {
+			auther = user.Authenticator()
+		}
+		if err != nil && err != errNoAuth {
+			return InternalError("%v", err)
+		}
```
Comment (niemeyer) [inline daemon/api.go]: This logic is all exactly the same between the two paths. Can we please transform this into:
```
auther, err := c.d.auther(req)
if err != nil {
        return err
}

**Pull request merged**